### PR TITLE
Structure_Engine: Add support for local orientation of Panels and FEMeshes, and Orient methods

### PR DESCRIPTION
**Pull request by IsakNaslundBh**
### NOTE: Depends on 

   
https://github.com/BHoM/BHoM/pull/974

### Issues addressed by this PR

Adding methods related to https://github.com/BHoM/BHoM/issues/963

Adding general method to support the addition of an orientationangle to Panel and FEMeshFace.
Centralising and aligning handling of orientation angle between bar and area element.
Some quality of life methods added.

### Test files

Versioning test files and testfiles for general functionality. Note that re-building versioning toolkit is needed to test the versioning files.:

https://burohappold.sharepoint.com/:f:/s/BHoM/EgtVAo-WZLxMuZ4smLsbhm8BzqRseNyNO8QxuZBWV6qqMw?e=ZglInM

### Changelog

- Adding method for getting local orientation from Panel and FEMeshFaces
- Adding method for extracting CoordinateSystem(s) from Panel and FEMesh(Faces)
- Adding methods for setting local orientation(s) for Panels and FEMeshes
- Adding localX vector input for Panel and FEMesh creates
- Adding method, centralising, calculation of orientation angle for Bar and Area elements
- Adding method for setting normal to a Bar
- Adding method for orienting an element towards a point for Bar, Panel and FEMesh(Faces)

### CODE REVIEW

File: Structure_Engine/Compute/OrientationAngle.cs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+{
+    public static partial class Compute
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Calculates the orientation angle of a Bar based on a normal vector and a centreline.")]
+        [Input("normal", "Vector to be used as normal of the Bar. This vector should generally be orthogonal to the Bar, if it is not, it will be made orthogonal by projecting it to the section plane of the Bar (a plane that has that Bar tangent as its normal). This means that the Normal cannot be paralell to the Tangent of the Bar. \n" +
+                         "Vector will be used to determain the orientation angle of the Bar. This is done by measuring the counter clockwise angle in the section plane of the Bar between a reference Vector and the provided Vector. For a non-vertical Bar, the reference vector will be the global Z-axis. For a vertical bar the reference vector will be a vector that is orthogonal to the tangent vector of the Bar and the global Y-axis.")]
```
Comment (peterjamesnugent) [inline Structure_Engine/Compute/OrientationAngle.cs:42]: determine

File: Structure_Engine/Compute/OrientationAngle.cs [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+            else
+            {
+                normal = normal.Normalise();
+                Vector tan = (centreline.End - centreline.Start).Normalise();
+
+                double dot = normal.DotProduct(tan);
+
+                if (Math.Abs(1 - dot) < Tolerance.Angle)
+                {
+                    Reflection.Compute.RecordError("The normal is parallell to the centreline of the Bar. Orientation angle could not be calcualted.");
```
Comment (peterjamesnugent) [inline Structure_Engine/Compute/OrientationAngle.cs:60]: parallel
OrientationAngle

File: Structure_Engine/Compute/OrientationAngle.cs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+                orientationAngle = reference.Angle(normal, new Plane { Normal = tan });
+            }
+
+            return orientationAngle;
+        }
+
+        /***************************************************/
+
+        [Description("Calculates the orientation angle of an area element based on the normal of the element and a provided local X direction.\n" +
+                     "The orientation angle is calculated as the angle between localX and a reference vector in the plane of the element, defined by the normal.\n" +
```
Comment (peterjamesnugent) [inline Structure_Engine/Compute/OrientationAngle.cs:86]: local x

File: Structure_Engine/Compute/OrientationAngle.cs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+                     "This reference vector is the global X-vector projected to the plane of the element, except for the case were the normal is parallel to the global X-axis." +
+                     "For this case the reference vector is instead the crossproduct of global Y and the normal, projected to the plane of the element.")]
```
Comment (peterjamesnugent) [inline Structure_Engine/Compute/OrientationAngle.cs:88]: cross product

File: Structure_Engine/Compute/OrientationAngle.cs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        [Output("orientationAngle", "The calculated orientation angle of the area element. Will return NaN if the normal and local x are parallel.")]
+        public static double OrientationAngleAreaElement(this Vector normal, Vector localX)
+        {
+            localX = localX.Normalise();
+            normal = normal.Normalise();
+            double dot = normal.DotProduct(localX);
+
+            if (Math.Abs(1 - dot) < Tolerance.Angle)
+            {
+                Reflection.Compute.RecordError("The provided localX is parallel to the normal of the element. The orientation angle could not be calculated.");
```
Comment (peterjamesnugent) [inline Structure_Engine/Compute/OrientationAngle.cs:100]: local x

File: Structure_Engine/Compute/OrientationAngle.cs [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+                return double.NaN;
+            }
+            else if (Math.Abs(dot) > Tolerance.Angle)
+            {
+                Reflection.Compute.RecordWarning("The provided localX is not in the Plane of the element and will get projected");
```
Comment (peterjamesnugent) [inline Structure_Engine/Compute/OrientationAngle.cs:105]: local x
plane

File: Structure_Engine/Compute/OrientationAngle.cs [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+            Vector refVec;
+
+            if (normal.IsParallel(Vector.XAxis) == 0)
+            {
+                //Normal is not paralell to the global X-axis
+                refVec = Vector.XAxis;
+            }
+            else
+            {
+                //Normal _is_ paralell to the global X-axis
```
Comment (peterjamesnugent) [inline Structure_Engine/Compute/OrientationAngle.cs:118]: parallel

File: Structure_Engine/Modify/OrientTowards.cs [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        {
+            Point centre = panel.Centroid();
+            Vector localX = (orientationPoint - centre).Normalise();
+
+            return panel.SetLocalOrientation(localX);
+        }
+
+        /***************************************************/
+
+        [Description("Orients the direction of the local x of each FEMeshFace of the FEMesh towards the provided point. The local x vectors is calculated as the vector from the centroid of each FEMeshFace of the FEMesh to the provided point.")]
```
Comment (peterjamesnugent) [inline Structure_Engine/Modify/OrientTowards.cs:70]: local x vectors are calculated

File: Structure_Engine/Modify/OrientTowards.cs [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        public static FEMesh OrientTowards(this FEMesh mesh, Point orientationPoint)
+        {
+            FEMesh clone = mesh.GetShallowClone() as FEMesh;
+            clone.Faces = clone.Faces.Select(x => x.OrientTowards(mesh, orientationPoint)).ToList();
+            return clone;
+        }
+
+        /***************************************************/
+
+        [Description("Orients the direction of the local x of the FEMeshFace towards the provided point. The local x vectors is calculated as the vector from the centroid of the FEMeshFace of to the provided point.")]
```
Comment (peterjamesnugent) [inline Structure_Engine/Modify/OrientTowards.cs:83]: local x vectors are calculated

File: Structure_Engine/Modify/SetNormal.cs [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+{
+    public static partial class Modify
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Aligns the normal of the bar with the provided normal by updating the OrientationAngle.")]
+        [Input("bar", "The Bar to update.")]
+        [Input("normal", "Vector to be used as normal of the Bar. This vector should generally be orthogonal to the Bar, if it is not, it will be made orthogonal by projecting it to the section plane of the Bar (a plane that has that Bar tangent as its normal). This means that the Normal cannot be paralell to the Tangent of the Bar. \n" +
```
Comment (peterjamesnugent) [inline Structure_Engine/Modify/SetNormal.cs:42]: parallel

File: Structure_Engine/Modify/SetNormal.cs [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+                         "Vector will be used to determain the orientation angle of the Bar. This is done by measuring the counter clockwise angle in the section plane of the Bar between a reference Vector and the provided Vector. For a non-vertical Bar, the reference vector will be the global Z-axis. For a vertical bar the reference vector will be a vector that is orthogonal to the tangent vector of the Bar and the global Y-axis.")]
```
Comment (peterjamesnugent) [inline Structure_Engine/Modify/SetNormal.cs:43]: determine

File: Structure_Engine/Query/CoordinateSystem.cs [hunk 12]
```diff
@@ -59,6 +60,39 @@ public static Cartesian CoordinateSystem(this Bar bar)
         }
 
         /***************************************************/
+
+        [Description("Get the carteseian coordinate system descibring the position and local orientation of the Panel in the global coordinate system where the z-axis is the normal of the Panel and the x- and y-axes are the directions of the local in-plane axes.")]
```
Comment (peterjamesnugent) [inline Structure_Engine/Query/CoordinateSystem.cs:64]: x and y axes

File: Structure_Engine/Query/CoordinateSystem.cs [hunk 13]
```diff
@@ -59,6 +60,39 @@ public static Cartesian CoordinateSystem(this Bar bar)
...
+        [Output("CoordinateSystem", "The local cartesian coordinate system of the Panel.")]
+        public static Cartesian CoordinateSystem(this Panel panel)
+        {
+            Basis orientation = panel.LocalOrientation();
+            return new Cartesian(panel.Centroid(), orientation.X, orientation.Y, orientation.Z);
+        }
+
+        /***************************************************/
+
+        [Description("Get the carteseian coordinate system descibring the position and local orientation of the FEMeshFaces of the FEMesh in the global coordinate system where the z-axis is the normal of the FEMeshFace and the x- and y-axes are the directions of the local in-plane axes.")]
```
Comment (peterjamesnugent) [inline Structure_Engine/Query/CoordinateSystem.cs:75]: x and y axes

File: Structure_Engine/Query/CoordinateSystem.cs [hunk 14]
```diff
@@ -59,6 +60,39 @@ public static Cartesian CoordinateSystem(this Bar bar)
...
+        [Input("mesh", "The FEMesh to extract the local coordinate systems of the FEMeshFaces from.")]
+        [Output("CoordinateSystem", "The local cartesian coordinate system of the FEMeshFaces of the FEMesh.")]
+        public static List<Cartesian> CoordinateSystem(this FEMesh mesh)
+        {
+            return mesh.Faces.Select(x => x.CoordinateSystem(mesh)).ToList();
+        }
+
+        /***************************************************/
+
+        [Description("Get the carteseian coordinate system descibring the position and local orientation of the FEMeshFace in the global coordinate system where the z-axis is the normal of the FEMeshFace and the x- and y-axes are the directions of the local in-plane axes.")]
```
Comment (peterjamesnugent) [inline Structure_Engine/Query/CoordinateSystem.cs:85]: x and y axes

File: Structure_Engine/Query/LocalOrientation.cs [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+
+namespace BH.Engine.Structure
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Get the Vector basis system descibring the local axis orientation of the Panel in the global coordinate system where the z-axis is the normal of the panel and the x- and y-axes are the directions of the local in-plane axes.")]
```
Comment (peterjamesnugent) [inline Structure_Engine/Query/LocalOrientation.cs:43]: x and y axes

File: Structure_Engine/Query/LocalOrientation.cs [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+        [Input("panel", "The Panel to extract the local orientation from from.")]
+        [Output("orienation", "The local orientation of the Panel as a vector Basis.")]
+        public static Basis LocalOrientation(this Panel panel)
+        {
+            return LocalOrientation(Engine.Spatial.Query.Normal(panel), panel.OrientationAngle);
+        }
+
+        /***************************************************/
+
+        [Description("Get the Vector basis system descibring the local axes orientation of the faces of the FEMesh in the global coordinate system where the z-axis is the normal of each face and the x- and y-axes are the directions of the local in-plane axes.")]
```
Comment (peterjamesnugent) [inline Structure_Engine/Query/LocalOrientation.cs:53]: x and y axes

File: Structure_Engine/Query/LocalOrientation.cs [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+        [Input("mesh", "The FEMesh to extract the local orientations from from.")]
+        [Output("orienations", "The local orientations of the mesh as a vector Basis.")]
+        public static List<Basis> LocalOrientations(this FEMesh mesh)
+        {
+            return mesh.Faces.Select(x => x.LocalOrientation(mesh)).ToList();
+        }
+
+        /***************************************************/
+
+        [Description("Get the Vector basis system descibring the local axes orientation of a face of a FEMesh in the global coordinate system where the z-axis is the normal of each face and the x- and y-axes are the directions of the local in-plane axes.")]
```
Comment (peterjamesnugent) [inline Structure_Engine/Query/LocalOrientation.cs:63]: x and y axes

**Review by peterjamesnugent — CHANGES_REQUESTED** [review]
Changes to method attributes.

File: Structure_Engine/Modify/OrientTowards.cs (hunk 9 above)
Comment (IsakNaslundBh) [reply Structure_Engine/Modify/OrientTowards.cs:83]: keeping 'is', but getting rid of s as single face

File: Structure_Engine/Query/CoordinateSystem.cs [hunk 18]
```diff
@@ -59,6 +60,39 @@ public static Cartesian CoordinateSystem(this Bar bar)
         }
 
         /***************************************************/
+
+        [Description("Get the carteseian coordinate system descibring the position and local orientation of the Panel in the global coordinate system where the z-axis is the normal of the Panel and the x and y axes are the directions of the local in-plane axes.")]
```
Comment (pawelbaran) [inline Structure_Engine/Query/CoordinateSystem.cs:64]:
```suggestion
        [Description("Get the Cartesian coordinate system describing the position and local orientation of the Panel in the global coordinate system where the z-axis is the normal of the Panel and the x and y axes are the directions of the local in-plane axes.")]
```

File: Structure_Engine/Query/CoordinateSystem.cs [hunk 19]
```diff
@@ -59,6 +60,39 @@ public static Cartesian CoordinateSystem(this Bar bar)
...
+        [Output("CoordinateSystem", "The local cartesian coordinate system of the Panel.")]
+        public static Cartesian CoordinateSystem(this Panel panel)
+        {
+            Basis orientation = panel.LocalOrientation();
+            return new Cartesian(panel.Centroid(), orientation.X, orientation.Y, orientation.Z);
+        }
+
+        /***************************************************/
+
+        [Description("Get the carteseian coordinate system descibring the position and local orientation of the FEMeshFaces of the FEMesh in the global coordinate system where the z-axis is the normal of the FEMeshFace and the x and y axes are the directions of the local in-plane axes.")]
```
Comment (pawelbaran) [inline Structure_Engine/Query/CoordinateSystem.cs:75]:
```suggestion
        [Description("Get the Cartesian coordinate system descibring the position and local orientation of the FEMeshFaces of the FEMesh in the global coordinate system where the z-axis is the normal of the FEMeshFace and the x and y axes are the directions of the local in-plane axes.")]
```

File: Structure_Engine/Query/CoordinateSystem.cs [hunk 20]
```diff
@@ -59,6 +60,39 @@ public static Cartesian CoordinateSystem(this Bar bar)
...
+        [Input("mesh", "The FEMesh to extract the local coordinate systems of the FEMeshFaces from.")]
+        [Output("CoordinateSystem", "The local cartesian coordinate system of the FEMeshFaces of the FEMesh.")]
+        public static List<Cartesian> CoordinateSystem(this FEMesh mesh)
+        {
+            return mesh.Faces.Select(x => x.CoordinateSystem(mesh)).ToList();
+        }
+
+        /***************************************************/
+
+        [Description("Get the carteseian coordinate system descibring the position and local orientation of the FEMeshFace in the global coordinate system where the z-axis is the normal of the FEMeshFace and the x and y axes are the directions of the local in-plane axes.")]
```
Comment (pawelbaran) [inline Structure_Engine/Query/CoordinateSystem.cs:85]:
```suggestion
        [Description("Get the Cartesian coordinate system descibring the position and local orientation of the FEMeshFace in the global coordinate system where the z-axis is the normal of the FEMeshFace and the x and y axes are the directions of the local in-plane axes.")]
```

File: Structure_Engine/Query/CoordinateSystem.cs [hunk 21]
```diff
@@ -59,6 +60,39 @@ public static Cartesian CoordinateSystem(this Bar bar)
         }
 
         /***************************************************/
+
+        [Description("Get the carteseian coordinate system descibring the position and local orientation of the Panel in the global coordinate system where the z-axis is the normal of the Panel and the x and y axes are the directions of the local in-plane axes.")]
+        [Input("panel", "The Panel to extract the local coordinate system from.")]
+        [Output("CoordinateSystem", "The local cartesian coordinate system of the Panel.")]
```
Comment (pawelbaran) [inline Structure_Engine/Query/CoordinateSystem.cs:66]:
```suggestion
        [Output("coordinateSystem", "Local Cartesian coordinate system of the Panel.")]
```

File: Structure_Engine/Query/CoordinateSystem.cs [hunk 22]
```diff
@@ -59,6 +60,39 @@ public static Cartesian CoordinateSystem(this Bar bar)
...
+        {
+            Basis orientation = panel.LocalOrientation();
+            return new Cartesian(panel.Centroid(), orientation.X, orientation.Y, orientation.Z);
+        }
+
+        /***************************************************/
+
+        [Description("Get the carteseian coordinate system descibring the position and local orientation of the FEMeshFaces of the FEMesh in the global coordinate system where the z-axis is the normal of the FEMeshFace and the x and y axes are the directions of the local in-plane axes.")]
+        [Input("mesh", "The FEMesh to extract the local coordinate systems of the FEMeshFaces from.")]
+        [Output("CoordinateSystem", "The local cartesian coordinate system of the FEMeshFaces of the FEMesh.")]
```
Comment (pawelbaran) [inline Structure_Engine/Query/CoordinateSystem.cs:77]:
```suggestion
        [Output("coordinateSystems", "Local Cartesian coordinate systems of the FEMeshFaces of the FEMesh.")]
```

File: Structure_Engine/Query/CoordinateSystem.cs [hunk 23]
```diff
@@ -59,6 +60,39 @@ public static Cartesian CoordinateSystem(this Bar bar)
...
+        {
+            return mesh.Faces.Select(x => x.CoordinateSystem(mesh)).ToList();
+        }
+
+        /***************************************************/
+
+        [Description("Get the carteseian coordinate system descibring the position and local orientation of the FEMeshFace in the global coordinate system where the z-axis is the normal of the FEMeshFace and the x and y axes are the directions of the local in-plane axes.")]
+        [Input("face", "The FEMeshFace to extract the local coordinate system from.")]
+        [Input("mesh", "The FEMesh to which the face belongs.")]
+        [Output("CoordinateSystem", "The local cartesian coordinate system of the FEMeshFace.")]
```
Comment (pawelbaran) [inline Structure_Engine/Query/CoordinateSystem.cs:88]:
```suggestion
        [Output("coordinateSystem", "Local Cartesian coordinate system of the FEMeshFace.")]
```

File: Structure_Engine/Query/LocalOrientation.cs [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+        public static Basis LocalOrientation(this Panel panel)
+        {
+            return LocalOrientation(Engine.Spatial.Query.Normal(panel), panel.OrientationAngle);
+        }
+
+        /***************************************************/
+
+        [Description("Get the Vector basis system descibring the local axes orientation of the faces of the FEMesh in the global coordinate system where the z-axis is the normal of each face and the x and y axes are the directions of the local in-plane axes.")]
+        [Input("mesh", "The FEMesh to extract the local orientations from from.")]
+        [Output("orienations", "The local orientations of the mesh as a vector Basis.")]
```
Comment (pawelbaran) [inline Structure_Engine/Query/LocalOrientation.cs:55]:
```suggestion
        [Output("orienations", "Local orientations of the mesh as a list of vector Bases.")]
```

File: Structure_Engine/Query/LocalOrientation.cs [hunk 25]
```diff
@@ -0,0 +1,106 @@
...
+
+        [Description("Get the Vector basis system descibring the local axes orientation of a face of a FEMesh in the global coordinate system where the z-axis is the normal of each face and the x and y axes are the directions of the local in-plane axes.")]
+        [Input("face", "The FEMeshFace to extract the local orientation from from.")]
+        [Input("mesh", "The FEMesh to which the face belongs.")]
+        [Output("orienation", "The local orientation of the FEMeshFace as a vector Basis.")]
+        public static Basis LocalOrientation(this FEMeshFace face, FEMesh mesh)
+        {
+            return LocalOrientation(face.Normal(mesh), face.OrientationAngle);
+        }
+
```
Comment (pawelbaran) [inline Structure_Engine/Query/LocalOrientation.cs:71]: two blanks 🙊

File: Structure_Engine/Query/Normal.cs [hunk 26]
```diff
@@ -46,6 +48,52 @@ public static Vector Normal(this Bar bar)
             return bar.Centreline().ElementNormal(bar.OrientationAngle);
         }
 
+        /***************************************************/
+
+        [Description("Returns the local z-axes of all FEMeshFaces in the FEMesh. Can only extract normals for 3 or 4-sided faces.")]
+        [Input("mesh", "The FEMesh to extract the face normals from.")]
+        [Output("normal", "Vector representing the local z-axis faces. List order coresponds to the order of the faces.")]
```
Comment (pawelbaran) [inline Structure_Engine/Query/Normal.cs:55]:
```suggestion
        [Output("normal", "List of vectors representing the local z-axes of mesh faces. List order corresponds to the order of the faces.")]
```

File: Structure_Engine/Query/Normal.cs [hunk 27]
```diff
@@ -46,6 +48,52 @@ public static Vector Normal(this Bar bar)
...
+        {
+            return mesh.Faces.Select(x => x.Normal(mesh)).ToList();
+        }
+
+        /***************************************************/
+
+        [Description("Returns the local z-axis of an FEMeshFace. Can only extract normals for 3 or 4-sided faces.")]
+        [Input("face", "The FEMeshFace to evaluate the normal of.")]
+        [Input("mesh", "The FEMesh to which the face belongs.")]
+        [Output("normal", "Vector representing the local z-axis element.")]
```
Comment (pawelbaran) [inline Structure_Engine/Query/Normal.cs:66]:
```suggestion
        [Output("normal", "Vector representing the local z-axis of a mesh face.")]
```

File: Structure_Engine/Query/Normal.cs [hunk 28]
```diff
@@ -46,6 +48,52 @@ public static Vector Normal(this Bar bar)
...
+                normal = Engine.Geometry.Query.CrossProduct(pB - pA, pC - pB);
+            else
+            {
+                Point pD = mesh.Nodes[face.NodeListIndices[3]].Position;
+                normal = (Engine.Geometry.Query.CrossProduct(pA - pD, pB - pA)) + (Engine.Geometry.Query.CrossProduct(pC - pB, pD - pC));
+            }
+
+            return normal.Normalise();
+        }
+
```
Comment (pawelbaran) [inline Structure_Engine/Query/Normal.cs:96]: two blanks?

File: Structure_Engine/Modify/SetNormal.cs [hunk 29]
```diff
@@ -0,0 +1,62 @@
...
+
+namespace BH.Engine.Structure
+{
+    public static partial class Modify
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Aligns the normal of the bar with the provided normal by updating the OrientationAngle.")]
```
Comment (pawelbaran) [inline Structure_Engine/Modify/SetNormal.cs:40]:
```suggestion
        [Description("Aligns the normal of the bar to the provided normal by updating the OrientationAngle of the former.")]
```

File: Structure_Engine/Modify/OrientTowards.cs [hunk 30]
```diff
@@ -0,0 +1,97 @@
...
+            return clone;
+        }
+
+        /***************************************************/
+
+        [Description("Orients the direction of the local x of the FEMeshFace towards the provided point. The local x vector is calculated as the vector from the centroid of the FEMeshFace of to the provided point.")]
+        [Input("face", "The FEMeshFace to update the local orientation of.")]
+        [Input("mesh", "The FEMesh to which the face belongs.")]
+        [Input("orientationPoint", "The point to orient the local x of the FEMeshFaces towards.")]
+        [Output("mesh", "FEMeshFaces with updated local orientation. If the calcualtion of the orientation angle fails, the unmodified FEMeshFace is returned.")]
```
Comment (pawelbaran) [inline Structure_Engine/Modify/OrientTowards.cs:87]:
```suggestion
        [Output("meshFace", "FEMeshFace with updated local orientation. If the calcualtion of the orientation angle fails, the unmodified FEMeshFace is returned.")]
```

File: Structure_Engine/Modify/OrientTowards.cs [hunk 31]
```diff
@@ -0,0 +1,97 @@
...
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Orients the normal of the Bar towards the provided point. The normal is calculated as the vector from the centre of the bar to the provided point.")]
+        [Input("bar", "The Bar to update the orientation of.")]
+        [Input("orientationPoint", "The point to orient the normal of the Bar towards.")]
+        [Output("bar", "Bar with updated orientation. If the calcualtion of the orientation angle fails, the unmodified bar is returned.")]
+        public static Bar OrientTowards(this Bar bar, Point orientationPoint)
+        {
+            Point centre = (bar.StartNode.Position + bar.EndNode.Position) / 2;
```
Comment (pawelbaran) [inline Structure_Engine/Modify/OrientTowards.cs:48]: Choice of bar's midpoint as a reference point is arbitrary and will impose a non-perpendicular vector warning in every case where `orientationPoint` is not exactly _above_ the midpoint of the bar. Personally I would prefer not raising the warning at all, regardless of the location of the `orientationPoint`.

**Review by pawelbaran — CHANGES_REQUESTED** [review]
I have requested a few minor changes - please see the code comments.

File: Structure_Engine/Modify/OrientTowards.cs (hunk 31 above)
Comment (IsakNaslundBh) [reply Structure_Engine/Modify/OrientTowards.cs]: After discussions offline, get your point. Have change to use the closest point on the finite bar as reference, should limit the warnings to when no line can be drawn from the bar to the point that is orthogonal to the bar

**Review by pawelbaran — APPROVED** [review]
LGTM

**Comment by IsakNaslundBh** [comment]
@pawelbaran I had to rebase update this branch due to conflict. Mind re-checkinging/approving?

**Review by pawelbaran — APPROVED** [review]
LGTM

**Pull request merged**